Mathlib/Topology/Algebra/Module/PointwiseConvergence.lean

```diff
@@ -0,0 +1,154 @@
+as a continuous linear map.
+* `ContinousLinearMap.toPointwiseConvergenceCLM`: The canonical map from `E →SL[σ] F` to
+`E →SLₚₜ[σ] F` as a continuous linear map. This is the statement that bounded convergence is
+stronger than pointwise convergence.
+* `PointwiseConvergenceCLM.equivWeakDual`: The continuous equivalence between `E →Lₚₜ[𝕜] 𝕜` and
+`WeakDual 𝕜 E`.
+
+## Main statements
+
+* `PointwiseConvergenceCLM.tendsto_iff_forall_tendsto`: In the topology of pointwise convergence,
+`a` converges to `a₀` iff for every `x : E` the map `a · x` converges to `a₀ x`.
+* `PointwiseConvergenceCLM.continuous_of_continuous_eval`: A map to `g : α → E →SLₚₜ[σ] F` is
+continuous if for every `x : E` the evaluation `g · x` is continuous.
+
+## Notation
+
+* `E →SLₚₜ[σ] F` is space of continuous linear maps equipped with pointwise convergence topology.
+
+-/
+
+/-! ### Topology of pointwise convergence -/
+
+variable {α ι : Type*} [TopologicalSpace α]
+variable {𝕜 𝕜₁ 𝕜₂ : Type*} [NormedField 𝕜] [NormedField 𝕜₁] [NormedField 𝕜₂]
+variable {σ : 𝕜₁ →+* 𝕜₂}
+variable {E F : Type*} [AddCommGroup E] [TopologicalSpace E]
+  [AddCommGroup F] [TopologicalSpace F] [TopologicalAddGroup F]
+  [Module 𝕜 E] [Module 𝕜 F] [Module 𝕜₁ E] [Module 𝕜₂ F]
+
+open Topology
+
+variable (σ E F) in
+/-- The space of continuous linear maps equipped with the topology of pointwise convergence,
+sometimes also called the *strong operator topology*. We avoid this terminology since so many other
+things share similar names, and using "pointwise convergence" in the name is more informative.
+
+This topology is also known as the weak*-topology in the case that `σ = RingHom.id 𝕜` and `F = 𝕜` -/
+@[reducible]
+def PointwiseConvergenceCLM := UniformConvergenceCLM σ F {s : Set E | Finite s}
+
+@[inherit_doc]
+notation:25 E " →SLₚₜ[" σ "] " F => PointwiseConvergenceCLM σ E F
+
+@[inherit_doc]
+notation:25 E " →Lₚₜ[" R "] " F => PointwiseConvergenceCLM (RingHom.id R) E F
+
+namespace PointwiseConvergenceCLM
+
+protected theorem hasBasis_nhds_zero_of_basis
+    {ι : Type*} {p : ι → Prop} {b : ι → Set F} (h : (𝓝 0 : Filter F).HasBasis p b) :
+    (𝓝 (0 : E →SLₚₜ[σ] F)).HasBasis (fun Si : Set E × ι => Finite Si.1 ∧ p Si.2)
+      fun Si => { f : E →SLₚₜ[σ] F | ∀ x ∈ Si.1, f x ∈ b Si.2 } :=
+  UniformConvergenceCLM.hasBasis_nhds_zero_of_basis σ F { S | Finite S }
+    ⟨∅, Set.finite_empty⟩ (directedOn_of_sup_mem fun _ _ => Set.Finite.union) h
+
+protected theorem hasBasis_nhds_zero [TopologicalSpace F] [TopologicalAddGroup F] :
+    (𝓝 (0 : E →SLₚₜ[σ] F)).HasBasis
+      (fun SV : Set E × Set F => Finite SV.1 ∧ SV.2 ∈ (𝓝 0 : Filter F))
+      fun SV => { f : E →SLₚₜ[σ] F | ∀ x ∈ SV.1, f x ∈ SV.2 } :=
+  PointwiseConvergenceCLM.hasBasis_nhds_zero_of_basis (𝓝 0).basis_sets
+
+/-- In the topology of pointwise convergence, `a` converges to `a₀` iff for every `x : E` the map
+`a · x` converges to `a₀ x`. -/
+theorem tendsto_iff_forall_tendsto {p : Filter ι} {a : ι → E →SLₚₜ[σ] F} {a₀ : E →SLₚₜ[σ] F} :
```

Comment:
I think you'll make your life easier by adding these two lemmas right after [UniformOnFun.uniformContinuous_restrict](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/UniformConvergenceTopology.html#UniformOnFun.uniformContinuous_restrict)
```lean
theorem uniformEmbedding_toFun_finite :
    UniformEmbedding (toFun _ : (α →ᵤ[{s | s.Finite}] β) → (α → β)) := by
  refine ⟨⟨?_⟩, Function.injective_id⟩
  simp_rw [Pi.uniformity, comap_iInf, comap_comap]
  refine HasBasis.ext (hasBasis_iInf' fun i ↦ (Filter.basis_sets _).comap _)
    (UniformOnFun.hasBasis_uniformity α β _ ⟨∅, finite_empty⟩
      (directedOn_of_sup_mem fun _ _ ↦ .union))
    (fun ⟨S, U⟩ ⟨hS, hU⟩ ↦ ⟨⟨S, ⋂ x ∈ S, U x⟩, ⟨⟨hS, biInter_mem hS |>.mpr hU⟩,
      fun f hf ↦ mem_iInter₂.mpr fun x hx ↦ mem_iInter₂.mp (hf x hx) x hx⟩⟩)
    (fun ⟨S, U⟩ ⟨hS, hU⟩ ↦ ⟨⟨S, fun _ ↦ U⟩, ⟨hS, fun _ _ ↦ hU⟩, fun f hf x hx ↦
      mem_iInter₂.mp hf x hx⟩)

theorem embedding_toFun_finite :
    Embedding (toFun _ : (α →ᵤ[{s | s.Finite}] β) → (α → β)) :=
  (uniformEmbedding_toFun_finite α β).embedding
```
and then adding to the current file
```lean
variable (σ E F) in
protected theorem embedding_coeFn : Embedding ((↑) : (E →SLₚₜ[σ] F) → (E → F)) :=
  let _: UniformSpace F := TopologicalAddGroup.toUniformSpace F
  have _ : UniformAddGroup F := comm_topologicalAddGroup_is_uniform
  (UniformOnFun.embedding_toFun_finite E F).comp (UniformConvergenceCLM.embedding_coeFn σ F _)
```

This and the corresponding result about `UniformConvergenceCLM` will allow you to use any piece of API about product topologies and uniform convergence topologies, which mean you wouldn't have to do any nontrivial proofs I think.

E.g this particular lemma would follow immediately from [Inducing.tendsto_nhds_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Maps.html#Inducing.tendsto_nhds_iff) and [tendsto_pi_nhds](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Constructions.html#tendsto_pi_nhds), the one about continuity from [Inducing.continuous_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Maps.html#Inducing.continuous_iff) and [continuous_pi_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Constructions.html#continuous_pi_iff), etc...